lld/COFF/InputFiles.cpp

```diff
@@ -660,10 +660,15 @@ std::optional<Symbol *> ObjFile::createDefined(
 
     if (prevailing) {
       SectionChunk *c = readSection(sectionNumber, def, getName());
-      sparseChunks[sectionNumber] = c;
-      c->sym = cast<DefinedRegular>(leader);
-      c->selection = selection;
-      cast<DefinedRegular>(leader)->data = &c->repl;
+      if (c) {
+        sparseChunks[sectionNumber] = c;
+        c->sym = cast<DefinedRegular>(leader);
+        c->selection = selection;
+        cast<DefinedRegular>(leader)->data = &c->repl;
+      } else {
+        sparseChunks[sectionNumber] = nullptr;
+        return nullptr;
+      }
     } else {
       sparseChunks[sectionNumber] = nullptr;
```

Comment:
I see three assignments to `sparseChunks[sectionNumber]`, can we reduce nesting and arrange the code so that we store a variable which is sometimes null?

For your code change, essentially wrap the `c->` expressions in the `if (c)` check.

Is the `return nullptr` necessary, or can we return `leader`?

Reply:
I guess @aganea's suggestion of just moving the `if (!c) return nullptr;` after assigning `sparseChunks[sectionNumber] = c;` would simplify it a bit as well.

We can't return `leader` (I tested that and spent an awful lot of time hunting down why things were screwed up); the `cast<DefinedRegular>(leader)->data = &c->repl;` assignment is essential - otherwise the `leader` symbol is left with a nullptr `data`, which is expected to be initialized before the end.

As we're returning differently (`leader` vs `nullptr`) in the two cases where we're assigning `nullptr` into `sparseChunks[sectionNumber]`, I don't think we very neatly can fold all the three cases together, but @aganea's suggestion might make it more palatable.

Reply:
I debugged this code too, a while bit ago for https://reviews.llvm.org/D157136 and let's say, the way data flows isn't simple (even if the input data format is simple). Lots of interactions until things converge into place. I always wondered if instead of imperative code for a linker, if a data graph would be easier to follow.